### ISSUE TITLE
Game freeze when controler is out of battery

### ISSUE THREAD

**Issue by raginglazor** [ndeadly/MissionControl#9]
The game freezed when the PS4 controler ran out of battery, happend twice, once in retroarch and once in carrion

**Comment by raginglazor** [comment]
Also the système is still running as i Can press home and close the game

**Comment by ndeadly** [comment]
I don't think I've ever had a switch controller die on me mid-game. Do you know what the usual behaviour is when this happens? Could it be that this is normal?

**Comment by raginglazor** [comment]
I was using a PS4 controler, had my joycon die mid game, but never got freezes like this, so it is certainly not a normal behaviour

**Comment by raginglazor** [comment]
The switch was docked when both freeze occured

**Comment by raginglazor** [comment]
When the controller is out of battery, the Switch prompt you with a controller low battery pop UP on the top left corner,  i tried to reconnect the controller which was out of battery ans then the game freezed

**Comment by ndeadly** [comment]
Please try to contain your information to as fewer messages as possible. Github issues are not a chat service.

I'll try to look into this, but it's something that may take me a while to replicate. It's possible my battery level mappings are off, leading to a jump straight to "empty" rather than hitting "low battery" first or something like that. Let me know if you are able to provide any additional info on the issue.

**Comment by evertonstz** [comment]
> I don't think I've ever had a switch controller die on me mid-game. Do you know what the usual behaviour is when this happens? Could it be that this is normal?

Happened to me a couple time, what happens is: you know when you put the switch to sleep while running a game and when you wake up the switch again it asks for you to press L+R in a controller (Stardew Valley for example does this, some games don't do this tho)? This is what you see when you run out of battery, this overlay asking for you to press L+R in a controller.

**Comment by ndeadly** [comment]
@raginglazor I've made a new build with some changes to the DS4 battery handling code. Install this and see if the low battery still causes the freeze.

[MissionControl-0.1.0-ds4-battery-fix.zip](https://github.com/ndeadly/MissionControl/files/5158457/MissionControl-0.1.0-ds4-battery-fix.zip)

**Comment by raginglazor** [comment]
I'll try it, and keep you updated

**Comment by fullmetal1** [comment]
I was having a similar issue.

My DS4 (2nd gen) was low battery last night so I left it to charge. I would try to connect it (pressing PS button), and after the indicator stopped flashing, instead of turning a solid blue, it would stay unlit, but the controller was connected, because trying to connect it again (PS button) wouldn't get a response. The Switch froze as well. (This happened a few times)

I installed your latest fix, and restarted my switch, and reconnected, and now it's connected fine and says that the battery was full.

Incidentally I also noticed it mentioning the battery was low more quickly than I would expect. I'll keep playing to let the battery discharge (I guess I'll also try to check battery level in windows), and I'll let you know if it seems better.

**Comment by ndeadly** [comment]
@fullmetal1 good to hear your issue has been resolved. Hopefully this solves the problems of the other users too.

Thanks for the feedback. If you could keep an eye on the battery level that would be great. The fix I made will stop the battery level value going out of bounds, but I have no idea how accurate the indicator actually is vs what you would expect from a normal Switch controller. Battery issues can be time consuming to debug so I haven't invested a lot of time in this to date.  I scale the incoming battery values the best I can to the range of the Switch format, but I'm starting to wonder whether it might be nonlinear.
The battery level values documented by dekuNukem [here](https://github.com/dekuNukem/Nintendo_Switch_Reverse_Engineering/blob/master/bluetooth_hid_notes.md#standard-input-report-format) are `8=full, 6=medium, 4=low, 2=critical, 0=empty. LSB=Charging`. As you can see, a value of 6 is already considered to be "medium"

**Comment by fullmetal1** [comment]
Yeah, I was getting 'low battery' indicators on the switch at around 10-12 hours of gameplay, when this controller is only lightly used and apparently lasts about 20 hours of continuous gameplay. 

If I had to guess, full = 100-80%, medium = 79-30%, low = 29-10%, critical = 9-3%, empty = 2-0%, or something to that effect. Where is the commit where you fixed this issue? I'm curious what the current battery code for DS4 looks like

**Comment by ndeadly** [comment]
@raginglazor can you confirm this is fixed with the 0.2.0 release?

**Comment by fullmetal1** [comment]
So I was away last week (didn't bring my ds4) but I am back and still checking. Unfortunately the battery lasts quite a while and I've exhausted my dailies in animal crossing. I've updated to ams 14.2 (was on 14.1), and updating from your test build to 0.2.0 from the homebrew app store. Everything still seems to be working fine. I'll let you know when the ds4 battery dies without crashing the switch.
